README.md
# rc-camera

---

React Camera Component.

Base on [Camera](https://github.com/pixedelic/Camera).


[![NPM version][npm-image]][npm-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]

[npm-image]: http://img.shields.io/npm/v/rc-camera.svg?style=flat-square
[npm-url]: http://npmjs.org/package/rc-camera
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/rc-camera.svg?style=flat-square
[download-url]: https://npmjs.org/package/rc-camera


## Feature

- support ie8,ie8+,chrome,firefox,safari(basic usage)

- support edge,chrome,firefox,safari,opera(some transform animations)



## install

[![rc-camera](https://nodei.co/npm/rc-camera.png)](https://npmjs.org/package/rc-camera)


## Usage

```jsx
import Camera from "rc-camera";
import "rc-camera/assets/index.css"

const imgList = [{ img: "", content: "" }];

ReactDOM.render(<Camera imgList={imgList} />, mountNode);
```

## API

### props

<table class="table table-bordered table-striped">
    <thead>
      <tr>
        <th style="width: 100px;">name</th>
        <th style="width: 100px;">type</th>
        <th style="width: 100px;">default</th>
        <th>description</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>width</td>
        <td>String</td>
        <td>100%</td>
        <td>Camera container width</td>
      </tr>
      <tr>
        <td>aspectRatio</td>
        <td>Number</td>
        <td>0.5</td>
        <td>The ratio of height / width. You can fit width and height. </td>
      </tr>
      <tr>
        <td>prefixCls</td>
        <td>String</td>
        <td>rc-camera</td>
        <td>PrefixCls of css </td>
      </tr>
      <tr>  
        <td>imgList</td>
        <td>Array</td>
        <td>[{}]</td>
        <td>Set imgs of Camera. </td>
      </tr>
      <tr>
        <td>current</td>
        <td>Number</td>
        <td>0</td>
        <td>Set the current img of Camera.</td>
      </tr>
      <tr>
        <td>slideOn</td>
        <td>String</td>
        <td>random</td>
        <td>The slide target of the animation: random prev next.</td>
      </tr>
      <tr>
        <td>animateType</td>
        <td>String</td>
        <td>random</td>
        <td>The animation type.</tde
      </tr>
     <tr>
        <td>duration</>
        <td>Number</td>
        <td>1000</td>
        <td>Animation duration(ms).</td>
      </tr>
      <tr>
        <td>easing</td>
        <td>String</td>
        <td>cubic-bezier(0.77, 0, 0.175, 1)</td>
        <td>Animation easing.</td>
      </tr>
      <tr>
        <td>contentBar</td>
        <td>(content) => React.ReactNode</td>
        <td></td>
        <td>ContentBar of Camera. You will get each img content when sliding.</td>
      </tr>
      <tr>
        <td>contentBarWrapStyle</td>
        <td>Obejct</td>
        <td>{}</td>
        <td>ContentBarWrap style. You can change wrapper background.</td>
      </tr>
      <tr>
        <td>prevBtn</td>
        <td>ReactNode</td>
        <td></td>
        <td>The prev Button of Camera.</td>
      </tr>
      <tr>
        <td>nextBtn</td>
        <td>ReactNode</td>
        <td></td>
        <td>The next Button of Camera.</td>
      </tr>
      <tr>
        <td>pagination</td>
        <td>Boolean</td>
        <td>true</td>
        <td>Pagination of Camera.</td>
      </tr>
      <tr>
        <td>autoPlay</td>
        <td>Boolean</td>
        <td>false</td>
        <td>Whether to auto play the Camera.</td>
      </tr>
      <tr>
        <td>autoPlayTime</td>
        <td>Number</td>
        <td>5000</td>
        <td>The interval time of autoPlay</td>
      </tr>
      <tr>
        <td>showThumbnail</td>
        <td>Boolean</td>
        <td>true</td>
        <td>Whether to show thumbnail when hover the pagination. </td>
      </tr>
      <tr>
        <td>onChange</td>
        <td>Function(current, nextIndex, isAuto)</td>
        <td></td>
        <td>Callback executed when slide to next img. </td>
      </tr>
      <tr>
        <td>onAnimationOver</td>
        <td>Function(current)</td>
        <td></td>
        <td>Callback executed when the animation is over. </td>
      </tr>
    </tbody>
</table>

## Development

```
npm install
npm start
```

## Example

http://localhost:8007/examples/default.html

online example: https://yhspehy.github.io/rc-camera/examples/

## License

rc-camera is released under the MIT license.


HISTORY.md
# History

----

## 1.0.3 / 2022-06-23

update deps

## 1.0.2 / 2019-06-06

add onChange and onAnimationOver hook

## 1.0.1 / 2019-05-29

add HISTORY.md and LICENSE.md

update READMD.md


src/Camera.js
import React from "react";
import PropTypes from "prop-types";

import TargetContainer from "./TargetContainer";
import ButtonContainer from "./ButtonContainer";
import Pagination from "./Pagination";
import animateTypeList, { CSS3Animation } from "./utils/config";

let autoPlayTimeOut;

const animationList = [...animateTypeList, ...CSS3Animation];

const animationNum = animationList.length;

class Camera extends React.PureComponent {
  static propTypes = {
    width: PropTypes.string,
    aspectRatio: PropTypes.number,
    prefixCls: PropTypes.string,
    imgList: PropTypes.arrayOf(PropTypes.object),
    current: PropTypes.number,
    slideOn: PropTypes.string,
    animateType: PropTypes.string,
    duration: PropTypes.number,
    easing: PropTypes.string,
    contentBar: PropTypes.func,
    contentBarWrapStyle: PropTypes.object,
    prevBtn: PropTypes.node,
    nextBtn: PropTypes.node,
    pagination: PropTypes.bool,
    autoPlay: PropTypes.bool,
    autoPlayTime: PropTypes.number,
    showThumbnail: PropTypes.bool,
    onChange: PropTypes.func,
    onAnimationOver: PropTypes.func
  };

  static defaultProps = {
    width: "100%",
    aspectRatio: 0.5,
    prefixCls: "rc-camera",
    imgList: [],
    current: 0,
    slideOn: "random",
    animateType: "random",
    duration: 1000,
    easing: "cubic-bezier(0.77, 0, 0.175, 1)",
    contentBarWrapStyle: {},
    pagination: true,
    autoPlay: false,
    autoPlayTime: 5000,
    showThumbnail: true
  };

  constructor(props) {
    super(props);

    let current = props.current;

    if (current + 1 > props.imgList.length) {
      // console.warn("current大于imgList长度，将默认显示第一张图片");
      current = 0;
    }

    const animateType =
      props.animateType === "random" ? "scrollTop" : props.animateType;

    const slideOn = props.slideOn === "random" ? "next" : props.slideOn;
    this.cameraRef = React.createRef();

    this.state = {
      current,
      nextIndex: 0,
      isAnimate: false,
      height: 0,
      isHover: false,
      animateType,
      slideOn
    };
  }

  componentDidMount() {
    this.getHeight();
    this.setAutoPlay();
  }

  getHeight = async () => {
    const { aspectRatio } = this.props;
    this.setState({
      height: this.cameraRef.current.clientWidth * aspectRatio
    });
  };

  getAnimateType = () => {
    const { animateType } = this.props;
    if (animateType === "random") {
      const randomIndex = Math.floor(Math.random() * animationNum);
      const _animateType = animationList[randomIndex];
      this.setState({
        animateType: _animateType
      });
    }
  };

  getSlideOn = () => {
    const { animateType } = this.props;

    if (animateType === "random") {
      this.setState({
        slideOn: Math.random() > 0.7 ? "prev" : "next"
      });
    }
  };

  setAutoPlay = () => {
    const { current } = this.state;
    const { autoPlay, imgList, autoPlayTime } = this.props;
    if (autoPlay) {
      autoPlayTimeOut = setTimeout(() => {
        const i = current === imgList.length - 1 ? 0 : current + 1;
        this.handleClick(i, true);
      }, autoPlayTime);
    }
  };

  mouseEnter = () => {
    this.setState({
      isHover: true
    });
  };

  mouseLeave = () => {
    this.setState({
      isHover: false
    });
  };

  // Each action to next img(include click and autoPlay)
  handleClick = (next, isAuto = false) => {
    const { isAnimate, current } = this.state;
    const { onChange } = this.props;
    if (isAnimate) {
      return;
    }
    if (!isAuto && autoPlayTimeOut) {
      clearTimeout(autoPlayTimeOut);
    }
    if (onChange) {
      onChange(current, next, isAuto);
    }
    this.getAnimateType();
    this.getSlideOn();
    this.setState({
      nextIndex: next,
      isAnimate: true
    });
  };

  // handle animation when it's over
  handleAnimate = (current, isAnimate) => {
    const { onAnimationOver } = this.props;
    this.setState(
      {
        current,
        isAnimate: !!isAnimate
      },
      () => {
        if (onAnimationOver) {
          onAnimationOver(current);
        }
        this.setAutoPlay();
      }
    );
  };

  render() {
    const {
      imgList,
      prefixCls,
      width,
      duration,
      easing,
      contentBar,
      contentBarWrapStyle,
      prevBtn,
      nextBtn,
      pagination,
      showThumbnail
    } = this.props;
    const {
      current,
      height,
      isHover,
      nextIndex,
      isAnimate,
      animateType,
      slideOn
    } = this.state;

    return (
      <div className={prefixCls}>
        <div
          className={`${prefixCls}-container`}
          onMouseEnter={this.mouseEnter}
          onMouseLeave={this.mouseLeave}
          ref={this.cameraRef}
          style={{
            width,
            height
          }}
        >
          {height > 0 ? (
            <React.Fragment>
              <TargetContainer
                prefixCls={prefixCls}
                width={this.cameraRef.current.clientWidth}
                height={height}
                imgList={imgList}
                current={current}
                nextIndex={nextIndex}
                isAnimate={isAnimate}
                handleAnimate={this.handleAnimate}
                slideOn={slideOn}
                animateType={animateType}
                duration={duration}
                easing={easing}
                contentBar={contentBar}
                contentBarWrapStyle={contentBarWrapStyle}
              />

              <ButtonContainer
                width={this.cameraRef.current.clientWidth}
                height={height}
                prefixCls={prefixCls}
                imgList={imgList}
                current={current}
                handleClick={this.handleClick}
                isHover={isHover}
                prevBtn={prevBtn}
                nextBtn={nextBtn}
              />

              {pagination && (
                <Pagination
                  prefixCls={prefixCls}
                  imgList={imgList}
                  current={current}
                  nextIndex={nextIndex}
                  handleClick={this.handleClick}
                  showThumbnail={showThumbnail}
                />
              )}
            </React.Fragment>
          ) : null}
        </div>
      </div>
    );
  }
}

export default Camera;
